src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs

```diff
@@ -41,11 +40,11 @@ private static uint UpdateX86(uint crc, ReadOnlySpan<byte> source)
             x1 ^= Vector128.CreateScalar((ulong)crc);
             Vector128<ulong> x0 = Vector128.Create(0x0154442bd4, 0x01c6e41596).AsUInt64(); // k1, k2
 
-            srcRef = ref Unsafe.Add(ref srcRef, X86BlockSize);
-            length -= X86BlockSize;
+            srcRef = ref Unsafe.Add(ref srcRef, Vector128<byte>.Count * 4);
+            length -= Vector128<byte>.Count * 4;
 
             // Parallel fold blocks of 64, if any.
-            while (length >= X86BlockSize)
+            while (length >= Vector128<byte>.Count * 4)
             {
                 x5 = Pclmulqdq.CarrylessMultiply(x1, x0, CarrylessMultiplyLower);
                 Vector128<ulong> x6 = Pclmulqdq.CarrylessMultiply(x2, x0, CarrylessMultiplyLower);
```

Comment:
If you abstract this out into a helper, you can also support it on Arm64 by using the "polynomial multiply" instructions. Believe you specifically want `PMULL` and `PMULL2` which correspond to `AdvSimd.PolynmoialMultiplyWideningLower/Upper`

> Polynomial Multiply Long. This instruction multiplies corresponding elements in the lower or upper half of the
vectors of the two source SIMD&FP registers, places the results in a vector, and writes the vector to the destination
SIMD&FP register. The destination vector elements are twice as long as the elements that are multiplied.

Reply:
I actually considered using PMULL and PMULL2 on ARM when I was writing this, but I didn't do so because:

- Some of this implementation feels like it might be affected by byte order. Probably addressable, but it was a concern.
- I couldn't find an equivalent of `Sse2.ShiftRightLogical128BitLane` on ARM, so we'd need to create a less performant equivalent (correct me if I'm just blind on this one)
- ARM has a built-in CRC32 intrinsic that uses the same polynomial, so I assumed (potentially incorrectly) that it would be a better choice. So I was thinking I'd come back next and add an ARM-specific implementation.

Reply:
I've done a bit more research on this, and it appears that the PMULL approach is more performant on modern ARM that supports it than the CRC32 intrinsic on larger buffers because it can operate on wider data sets. I see evidence of commits in the Linux kernel and Java based on this. It seems like they use the CRC32 intrinsic as a fallback when PMULL is not available.

Based on this, I'm considering refactoring this to support ARM PMULL as well. However, I'm a bit stumped on how to test it since my dev laptop is Intel. Are there any tricks documented to test using QEMU or something similar?

Reply:
The easiest way is likely just to let our CI cover it if you don't have your own box.

I wouldn't expect any significant differences and you can filter on `BitConverter.IsLittleEndian` to ensure it works on BigEndian platforms if that's a concern.